In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats as stats
import seaborn as sns
import numpy as np
import warnings
warnings.filterwarnings("ignore")
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, MinMaxScaler
from sklearn.neighbors import LocalOutlierFactor
from sklearn.ensemble import IsolationForest
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import LabelEncoder
from scipy.stats import skew

In [35]:
df=pd.read_csv("/content/data_car.csv")

In [36]:
df.head(10)

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   
5  BMW    1 Series  2012  premium unleaded (required)      230.0   
6  BMW    1 Series  2012  premium unleaded (required)      300.0   
7  BMW    1 Series  2012  premium unleaded (required)      300.0   
8  BMW    1 Series  2012  premium unleaded (required)      230.0   
9  BMW    1 Series  2013  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   
5               6.0            MANUAL  rear wheel drive              2.0   
6               6.0            MANUAL  rear wheel drive              2.0   
7               6.0            MANUAL  rear wheel drive              2.0   
8               6.0            MANUAL  rear wheel drive              2.0   
9               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   
5                     Luxury,Performance      Compact         Coupe   
6                     Luxury,Performance      Compact   Convertible   
7                Luxury,High-Performance      Compact         Coupe   
8                                 Luxury      Compact   Convertible   
9                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500  
5           28        18        3916  31200  
6           26        17        3916  44100  
7           28        20        3916  39300  
8           28        18        3916  36900  
9           27        18        3916  37200

In [37]:
# Data exploration
print(df.shape) # Check number of rows and columns
print(df.columns) # Check column names
print(df.head()) # Check first few rows of data
print(df.info()) # Check data types and non-null values
print(df.describe()) # Check summary statistics

(11914, 16)
Index(['Make', 'Model', 'Year', 'Engine Fuel Type', 'Engine HP',
       'Engine Cylinders', 'Transmission Type', 'Driven_Wheels',
       'Number of Doors', 'Market Category', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP'],
      dtype='object')
  Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive         

**A) EXPLORATORY DATA ANALYSIS**

In [38]:
# Check number of rows and columns
print(df.shape)

(11914, 16)


In this dataset we have 11914 rows and 16 columns

In [39]:
df.columns

Index(['Make', 'Model', 'Year', 'Engine Fuel Type', 'Engine HP',
       'Engine Cylinders', 'Transmission Type', 'Driven_Wheels',
       'Number of Doors', 'Market Category', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP'],
      dtype='object')

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

From the above summary of a DataFrame's basic information, we could see that majority of the datatypes are of 'object' type. There are 5 integers [Year,Highway MPG,City MPG,MSRP and Popularity] and 3 Float types columns

In [41]:
#summary statistics
print(df.describe())


               Year    Engine HP  Engine Cylinders  Number of Doors  \
count  11914.000000  11845.00000      11884.000000     11908.000000   
mean    2010.384338    249.38607          5.628829         3.436093   
std        7.579740    109.19187          1.780559         0.881315   
min     1990.000000     55.00000          0.000000         2.000000   
25%     2007.000000    170.00000          4.000000         2.000000   
50%     2015.000000    227.00000          6.000000         4.000000   
75%     2016.000000    300.00000          6.000000         4.000000   
max     2017.000000   1001.00000         16.000000         4.000000   

        highway MPG      city mpg    Popularity          MSRP  
count  11914.000000  11914.000000  11914.000000  1.191400e+04  
mean      26.637485     19.733255   1554.911197  4.059474e+04  
std        8.863001      8.987798   1441.855347  6.010910e+04  
min       12.000000      7.000000      2.000000  2.000000e+03  
25%       22.000000     16.000000    549

In [42]:
#checking null values in each column
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

 Only 11 columns out of 16 has zero null values.

**B) DATA PRE-PROCESSING**

Missing Value Handling , Data Encoding , Feature Scaling , Handling Outliers , Feature Engineering

1. MARKET CATEGORY

MISSING VALUE HANDLING

In [43]:
mode = df['Market Category'].mode()[0]
df['Market Category'].fillna(mode, inplace=True)

In [44]:
#missing values
df['Market Category'].isna().sum()

0

For handling missing values in categorical variables, the mode is the most appropriate measure of central tendency to use. This is because categorical variables do not have a numeric scale, so we cannot use the mean or median as we do with continuous variables

DATA ENCODING

In [45]:
# Using pandas.get_dummies()
encoded_df = pd.get_dummies(df, columns=['Market Category'])

# Using scikit-learn's OneHotEncoder
from sklearn.preprocessing import OneHotEncoder
encoder = OneHotEncoder(sparse=False)
encoded_data = encoder.fit_transform(df[['Market Category']])


The "Market Category" column has been encoded using one-hot encoding,represents each unique category in the column as a binary feature column.It allows to represent categorical information in a format that is suitable for algorithms.

OUTLIER DETECTION

Outlier detection is typically applied to numerical data to identify extreme values that are significantly different from other values in the data. For categorical data, such as the "Market Category" column in this case, outlier detection is not applicable as there are no numerical values to compare.



2. ENGINE FUEL TYPE

MISSING VALUE HANDLING

In [46]:
mode = df['Engine Fuel Type'].mode()[0]
df['Engine Fuel Type'].fillna(mode, inplace=True)

In [47]:
#missing values
df['Engine Fuel Type'].isna().sum()

0

For handling missing values in categorical variables, the mode is the most appropriate measure of central tendency to use. This is because categorical variables do not have a numeric scale.

DATA ENCODING

In [48]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()
df['Engine Fuel Type'] = label_encoder.fit_transform(df['Engine Fuel Type'])


Label encoding was done by the column "Engine Fuel Type" which assigns a unique numeric label to each category in the variable.

OUTLIER DETECTION

Outlier detection is typically not applicable or meaningful for categorical variables. 

3. ENGINE HP

MISSING VALUE HANDLING

In [49]:
df['Engine HP'].fillna(0,inplace=True)

In [50]:
df['Engine HP'].isna().sum()

0

We filled null value of this particular column 'Engine HP' with '0' for dealing null values.

DATA ENCODING

Data encoding is not mandatory for numerical variables because numerical variables are already in a suitable format for many machine learning algorithms.

OUTLIER DETECTION

In [51]:
#outlier - zscore
z_scores = stats.zscore(df['Engine HP'])
threshold = 3
outlier_indices = np.where(z_scores > threshold)[0]
outlier_indices

array([  469,   473,   612,   613,   615,   617,   618,   619,   620,
         621,   622,   696,   698,   700,   701,   703,   704,   966,
         974,  1387,  1388,  1397,  1620,  1621,  1622,  1623,  1624,
        1625,  1626,  1627,  1628,  1629,  1630,  1631,  2172,  2182,
        2200,  2202,  2339,  2340,  2452,  2463,  2470,  2481,  2489,
        2500,  2675,  2679,  2683,  2849,  2850,  2851,  2855,  2856,
        2857,  2862,  2863,  2873,  2874,  2875,  2877,  2878,  2879,
        2886,  2887,  2888,  2889,  2890,  2970,  2974,  2980,  2981,
        2982,  2983,  2986,  2987,  2990,  2991,  2997,  2999,  3004,
        3008,  3014,  3021,  3022,  3023,  3219,  3220,  4024,  4644,
        4645,  4646,  4666,  4667,  4668,  4771,  4772,  4775,  5073,
        5597,  5599,  5774,  5775,  5776,  5777,  6350,  6351,  6998,
        6999,  7055,  7056,  7057,  7058,  7059,  7060,  7061,  7062,
        7063,  7992,  8486,  8822,  8823,  8826,  8833,  8839,  8840,
        8946,  8947,

4. ENGINE CYLINDERS

MISSING VALUE HANDLING

In [52]:
df['Engine Cylinders'].fillna(0,inplace=True)

In [53]:
df['Engine Cylinders'].isna().sum()

0

We filled null value of this particular column 'Engine Cylinders' with '0' for dealing null values.

DATA ENCODING

It is not necessary to do data encoding for numerical variables column.

OUTLIER DETECTION

In [54]:
Q1 = df['Engine Cylinders'].quantile(0.25)
Q3 = df['Engine Cylinders'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

potential_outliers = df[(df['Engine Cylinders'] < lower_bound) | (df['Engine Cylinders'] > upper_bound)]

# Find the lower and upper bounds
lower_bound = Q1 - 1.5*IQR
upper_bound = Q3 + 1.5*IQR
print("Upper bound:", upper_bound)
print("Lower bound:", lower_bound)


Upper bound: 9.0
Lower bound: 1.0


5. NUMBER OF DOORS

MISSING VALUE HANDLING

In [55]:
df['Number of Doors'].fillna(0,inplace=True)

In [56]:
df['Number of Doors'].isna().sum()

0

We filled null value of this particular column 'No. of Doors' with '0' for dealing null values.

DATA ENCODING

In [57]:
df['Number of Doors'] = df['Number of Doors'].map({'2 Doors': 2, '3 Doors': 3, '4 Doors': 4, '5 Doors': 5})


OUTLIER DETECTION

The "Number of Doors" column in the cars dataset represents a categorical variable that indicates the number of doors for each car. Since this column contains categorical data, traditional outlier detection methods are not applicable.

6. MAKE

DATA ENCODING

In [58]:
# Create a LabelEncoder object
le = LabelEncoder()
le

LabelEncoder()

In [59]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()
df['Make'] = label_encoder.fit_transform(df['Make'])


The LabelEncoder object is used to encode the "Make" column into numerical.The encoded data is then stored in a new column called "Make Code".

7. MODEL

In [60]:
#label encoding - 

leD = LabelEncoder()
df['Model Encoded'] = leD.fit_transform(df['Model'])

8. YEAR

MISSING VALUE HANDLING

In [61]:
#missing values
df['Year'].isna().sum()

0

There is no missing values.

Since encoding might give large number of additional columns and we ended to keep this feature as numerical itself, encoding is not ideal.

And the goal is to predict the price of a car
based on various features and attributes using the year column, then it is not necessary to remove outliers from the year column.

9. TRANSMISSION TYPE

In [62]:
# Encoding
one_hot_encoded = pd.get_dummies(df['Transmission Type'], prefix='Transmission')

# Using scikit-learn
from sklearn.preprocessing import OneHotEncoder
encoder = OneHotEncoder()
one_hot_encoded = encoder.fit_transform(df[['Transmission Type']])


10. DRIVEN WHEELS

In [63]:
#Encoding using One Hot Encoder
df_encoded = pd.get_dummies(df, columns=['Driven_Wheels'], prefix='Driven')


11. VEHICLE SIZE

In [65]:
#ENCODING BY ONE HOT ENCODING
encoded_df = pd.get_dummies(df, columns=['Vehicle Size'], prefix=['Size'])


This will create new binary columns for each category in the "vehicle size" column, with a prefix of "size" followed by the category name. The presence of a category in a row will be represented by a value of 1, and its absence will be represented by a value of 0.

12. VEHICLE STYLE

In [66]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()
df['Vehicle Style'] = label_encoder.fit_transform(df['Vehicle Style'])


13. HIGHWAY MPG

In [68]:
#outlier - zscore
z_scores = stats.zscore(df['highway MPG'])
threshold = 3
outlier_indices = np.where(z_scores > threshold)[0]
outlier_indices

array([ 539,  540,  541, 1119, 1680, 1681, 1682, 1983, 1984, 3716, 3717,
       3718, 3719, 4705, 4706, 4785, 4789, 4798, 5778, 5779, 5780, 5790,
       5791, 5792, 5793, 6385, 6386, 6387, 6388, 6389, 6390, 6391, 6392,
       6393, 6394, 6921, 6922, 6923, 6924, 6925, 6926, 6927, 6928, 6929,
       6930, 6931, 6932, 6933, 6934, 6935, 6936, 6937, 6938, 8373, 8374,
       8375, 9850, 9851, 9852, 9853, 9854, 9867, 9868, 9869, 9870, 9871,
       9872])

14. CITY MPG

In [70]:
#outlier - zscore
z_scores = stats.zscore(df['city mpg'])
threshold = 3
outlier_indices = np.where(z_scores > threshold)[0]
outlier_indices

array([ 539,  540,  541, 1202, 1203, 1204, 1205, 1206, 1207, 1208, 1209,
       1210, 1211, 1680, 1681, 1682, 1983, 1984, 3716, 3717, 3718, 3719,
       4705, 4706, 4785, 4789, 4798, 5778, 5779, 5780, 5790, 5791, 5792,
       5793, 6385, 6386, 6387, 6388, 6389, 6390, 6391, 6392, 6393, 6394,
       6680, 6686, 6921, 6922, 6923, 6924, 6925, 6926, 6927, 6928, 6929,
       6930, 6931, 6932, 6933, 6934, 6935, 6936, 6937, 6938, 7663, 7664,
       7665, 7666, 7667, 7668, 7669, 7670, 7671, 7672, 7673, 7674, 7675,
       7676, 7677, 7678, 7691, 7692, 7693, 7694, 7695, 7696, 7697, 7698,
       7699, 7700, 7701, 7702, 7703, 7704, 7705, 7706, 7707, 7708, 8373,
       8374, 8375, 9850, 9851, 9852, 9853, 9854, 9867, 9868, 9869, 9870,
       9871, 9872])

15. POPULARITY

In [71]:
Q1 = df['Popularity'].quantile(0.25)
Q3 = df['Popularity'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

potential_outliers = df[(df['Popularity'] < lower_bound) | (df['Popularity'] > upper_bound)]

# Find the lower and upper bounds
lower_bound = Q1 - 1.5*IQR
upper_bound = Q3 + 1.5*IQR
print("Upper bound:", upper_bound)
print("Lower bound:", lower_bound)

Upper bound: 4199.0
Lower bound: -1641.0


16. MSRP

In [72]:
#outlier - zscore
z_scores = stats.zscore(df['MSRP'])
threshold = 3
outlier_indices = np.where(z_scores > threshold)[0]
outlier_indices

array([  460,   462,   463,   464,   465,   466,   467,   468,   469,
         470,   471,   472,   473,   598,   606,   608,   609,   611,
         612,   613,   614,   615,   616,   617,   618,   619,   620,
         621,   622,   682,   683,   684,   696,   697,   698,   699,
         700,   701,   702,   703,   704,  1452,  1453,  1454,  1455,
        1456,  1457,  1458,  1459,  1460,  1461,  1620,  1621,  1622,
        1623,  1624,  1625,  1626,  1627,  1628,  1629,  1630,  1631,
        1676,  1677,  1678,  1679,  2040,  2041,  2339,  2340,  2855,
        2858,  2862,  2863,  2868,  2870,  2873,  2874,  2877,  2878,
        2879,  2886,  2887,  2888,  2889,  2890,  2891,  2892,  2893,
        2894,  2895,  2896,  2921,  3505,  3524,  3535,  3536,  3537,
        3538,  3539,  3540,  3541,  3542,  3543,  3544,  3545,  3546,
        3547,  3548,  3549,  3550,  3565,  4024,  4644,  4645,  4646,
        4649,  4654,  4659,  4666,  4667,  4668,  4775,  5216,  5217,
        5219,  5223,

In [73]:
!pip install category_encoders

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.9/81.9 kB 2.9 MB/s eta 0:00:00


In [74]:
#encoding - binary encoder

import category_encoders as ce

# Create an instance of the BinaryEncoder
Make_binary_encoder = ce.BinaryEncoder(cols=['Make'])

Make_binary = Make_binary_encoder.fit_transform(df['Make'])
df = pd.concat([df, Make_binary], axis=1)

In [75]:
Make_binary

Make_0  Make_1  Make_2  Make_3  Make_4  Make_5
0           0       0       0       0       0       1
1           0       0       0       0       0       1
2           0       0       0       0       0       1
3           0       0       0       0       0       1
4           0       0       0       0       0       1
...       ...     ...     ...     ...     ...     ...
11909       1       0       0       1       0       0
11910       1       0       0       1       0       0
11911       1       0       0       1       0       0
11912       1       0       0       1       0       0
11913       0       1       1       1       1       1

[11914 rows x 6 columns]

STANDARDIZATION AND NORMALIZATION

In [78]:
#choosing standardisation
# Standardize the ENGINE HP column
scaler = StandardScaler()
Engine_HP_standardized = scaler.fit_transform(df['Engine HP'].values.reshape(-1, 1))

# Convert the standardized column back to a pandas Series
Engine_HP_standardized = pd.Series(Engine_HP_standardized.ravel())

# Replace the original 
df['Engine HP'] = Engine_HP_standardized

# Save the updated dataframe to a new CSV file
df.to_csv('std_norm_dataset.csv', index=False)

In [79]:
#choosing standardisation
# Standardize the ENGINE CYLINDERS column
scaler = StandardScaler()
Engine_Cylinders_standardized = scaler.fit_transform(df['Engine Cylinders'].values.reshape(-1, 1))

# Convert the standardized column back to a pandas Series
Engine_Cylinders_standardized = pd.Series(Engine_Cylinders_standardized.ravel())

# Replace the original 
df['Engine Cylinders'] = Engine_Cylinders_standardized

# Save the updated dataframe to a new CSV file
df.to_csv('std_norm_dataset.csv', index=False)

In [80]:
#choosing standardisation
# Standardize the MSRP column
scaler = StandardScaler()
MSRP_standardized = scaler.fit_transform(df['MSRP'].values.reshape(-1, 1))

# Convert the standardized column back to a pandas Series
MSRP_standardized = pd.Series(MSRP_standardized.ravel())

# Replace the original 
df['MSRP'] = MSRP_standardized

# Save the updated dataframe to a new CSV file
df.to_csv('std_norm_dataset.csv', index=False)

In [81]:
#choosing standardisation
# Standardize the POPULARITY column
scaler = StandardScaler()
Popularity_standardized = scaler.fit_transform(df['Popularity'].values.reshape(-1, 1))

# Convert the standardized column back to a pandas Series
Popularity_standardized = pd.Series(Popularity_standardized.ravel())

# Replace the original 
df['Popularity'] = Popularity_standardized

# Save the updated dataframe to a new CSV file
df.to_csv('std_norm_dataset.csv', index=False)